// @bundle-stats/utils/lib-esm/webpack/types.js
export const METRIC_TOTALS_SUFIX_ALL = 'ALL';
export const METRIC_TOTALS_PREFIX = 'totalSizeByType';
export var Metric;
(function (Metric) {
    Metric["BUNDLE_SIZE"] = "totalSizeByTypeALL";
    Metric["INITIAL_SIZE_JS"] = "totalInitialSizeJS";
    Metric["INITIAL_SIZE_CSS"] = "totalInitialSizeCSS";
    Metric["CACHE_INVALIDATION"] = "cacheInvalidation";
    Metric["ASSET_COUNT"] = "assetCount";
    Metric["CHUNK_COUNT"] = "chunkCount";
    Metric["MODULE_COUNT"] = "moduleCount";
    Metric["DUPLICATE_MODULES_COUNT"] = "duplicateModulesCount";
    Metric["DUPLICATE_CODE"] = "duplicateCode";
    Metric["PACKAGE_COUNT"] = "packageCount";
    Metric["DUPLICATE_PACKAGES_COUNT"] = "duplicatePackagesCount";
    Metric["TOTAL_SIZE_JS"] = "totalSizeByTypeJS";
    Metric["TOTAL_SIZE_CSS"] = "totalSizeByTypeCSS";
    Metric["TOTAL_SIZE_IMG"] = "totalSizeByTypeIMG";
    Metric["TOTAL_SIZE_MEDIA"] = "totalSizeByTypeMEDIA";
    Metric["TOTAL_SIZE_FONT"] = "totalSizeByTypeFONT";
    Metric["TOTAL_SIZE_HTML"] = "totalSizeByTypeHTML";
    Metric["TOTAL_SIZE_OTHER"] = "totalSizeByTypeOTHER";
})(Metric || (Metric = {}));
