// use-query-params/adapters/reach/index.cjs.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  ReachAdapter: () => ReachAdapter
});
module.exports = __toCommonJS(src_exports);
var import_router = require("@reach/router");
var import_react = require("react");
function makeAdapter() {
  const adapter = {
    replace(location) {
      import_router.globalHistory.navigate(location.search || "?", {
        replace: true,
        state: location.state
      });
    },
    push(location) {
      import_router.globalHistory.navigate(location.search || "?", {
        replace: false,
        state: location.state
      });
    },
    get location() {
      return import_router.globalHistory.location;
    }
  };
  return adapter;
}
var ReachAdapter = ({ children }) => {
  const [adapter] = (0, import_react.useState)(makeAdapter);
  return children(adapter);
};
